client/src/components/File.js

```diff
@@ -2,225 +2,143 @@
-        let source = new MediaSource();
-        this.refs.video_tag.src = window.URL.createObjectURL(source);
+  get isAudio() {
+    return this.ext === 'mp3' || this.ext === 'ogg';
+  }
 
-        source.addEventListener('sourceopen', (e) => {
-          let sourceBuffer = source.addSourceBuffer(mimeCodec);
-          MessageStore.orbit.ipfs.files.cat(this.state.file, (err, res) => {
-            let buf = [];
-            sourceBuffer.addEventListener('updateend', () => {
-              if(buf.length > 0)
-                sourceBuffer.appendBuffer(buf.shift())
-            });
-            res
-              .on('error', (err) => console.error(err))
-              .on('data', (data) => {
-                if(!sourceBuffer.updating)
-                  sourceBuffer.appendBuffer(toArrayBuffer(data));
-                else
-                  buf.push(toArrayBuffer(data));
-              })
-              .on('end', () => {
-                console.log("END!!!")
-                setTimeout(() => {
-                  source.endOfStream()
-                  this.refs.video_tag.play();
-                }, 100);
-              })
-          });
-        }, false);
-      }, 500);
-    }
+  get isImage() {
+    return this.ext === 'png' || this.ext === 'jpg' || this.ext === 'gif';
+  }
 
-    this.setState({ showPreview: !this.state.showPreview });
+  get isHighlightable() {
+    return highlight.getLanguage(this.ext);
   }
 
-  _isTextFile(name) {
-    return highlight.getLanguage(this.state.name.split('.').pop());
+  handleClick(name) {
+    this.setState({
+      showPreview: !this.state.showPreview,
+      previewContent: 'Loading...',
+    }, () => {
+      if (this.state.showPreview) {
+        ChannelActions.loadFile(this.props.hash, blob => {
+          let previewContent = 'Unable to display file.';
+          if (blob) {
+            const url = window.URL.createObjectURL(blob);
+            if (this.isAudio) {
+              previewContent = <audio height={200} src={url} controls autoPlay={true} />;
+            } else if (this.isImage) {
+              previewContent = <img height={200} src={url} />;
+            } else if (this.isVideo) {
+              previewContent = <video height={200} src={url} controls autoPlay={false} />;;
+            } else {
+              var fileReader = new FileReader();
+              fileReader.onload = (event) => {
+                const text = readUTF8String(new Uint8Array(event.target.result));
+                if (this.isHighlightable) {
+                  previewContent = <Highlight>{text}</Highlight>;
+                } else {
+                  previewContent = text;
+                }
+                this.setState({ previewContent });
+              };
+              fileReader.readAsArrayBuffer(blob);
+              return;
+            }
+          }
+          this.setState({ previewContent });
+        });
+      }
+    });
   }
```

Comment:
We need to put the MediaSource API based video player back in here. Atm, it's the only way to play a video from ipfs as a stream. See below for more info on catting a file. The main point here is that the file is a *stream*, not pre-buffered.

We could separate the video player to its own component?